src/v/kafka/server/connection_context.cc

```diff
@@ -186,32 +186,71 @@ bool connection_context::is_finished_parsing() const {
     return conn->input().eof() || _server.abort_requested();
 }
 
+connection_context::delay_t connection_context::calculate_throttle_delays(
+  const request_header& hdr, const size_t request_size) {
+    using clock = quota_manager::clock;
+    static_assert(std::is_same_v<clock, delay_t::clock>);
+
+    // Throttle on client based quotas
+    const auto client_quota_delay = _server.quota_mgr().record_tp_and_throttle(
+      hdr.client_id, request_size);
+
+    // Throttle on shard wide quotas
+    const auto now = clock::now();
+    _server.quota_mgr().record_request_tp(request_size, now);
+    const clock::duration shard_delay_enforce = now >= _throttled_until
+                                                  ? clock::duration::zero()
+                                                  : _throttled_until - now;
+    const clock::duration shard_delay = _server.quota_mgr().get_shard_delay();
+    _throttled_until = now + shard_delay;
+    const auto shard_delay_request = shard_delay - shard_delay_enforce;
+
+    // Sum up
+    const clock::duration delay_enforce = std::max(
```

Comment:
Is there a maximum amount we could throttle? There already exist a config option for the max value we'd ever throttle clients for.

Reply:
There is, and it is enforced by each throttle delay source (client and shard) separately. It may make sense to pop that limit here at this aggregate level instead though.